### ISSUE TITLE
[Extension types] Failure to recognize constant results of extension type constructors in switches

### ISSUE THREAD

**Issue by lrhn** [dart-lang/sdk#53932]
Example:
```dart
enum Bool {
  _true._(true),
  _false._(false);
  final bool isTrue;
  const Bool._(this.isTrue);
  const factory Bool(bool value) = _BoolHelper;
}

extension type const _BoolHelper._(Bool _) implements Bool {
  const _BoolHelper(bool isTrue) : this._(isTrue ? Bool._true : Bool._false);
}

void main() {
  var b = Bool(DateTime.now().millisecondsSinceEpoch > 0); // True.
  var ago = switch (b) {
    const Bool(false) => "Some time ago",
    const Bool(true) => "More recent",
  };
  print(ago); // More recent
}
```
This code runs in CFE and prints "More recent".

The analyzer reports the following:
```
  error - switch-const-value-fail.dart:15:13 - The type 'Bool' is not
          exhaustively matched by the switch cases since it doesn't
          match 'Bool._true'. Try adding a wildcard pattern or cases
          that match 'Bool._true'. - non_exhaustive_switch_expression
```
Since the actual two constant values of the switch cases are indeed `Bool._false` and `Bool._true` (it can't be any other value, since it's an `enum`, and it's not the same value), the switch should be exhaustive.

The front-end recognizes the actual constants from evaluating `const Bool(false)` and `const Bool(true)`.

**Comment by bwilkerson** [comment]
@kallentu

**Comment by kallentu** [comment]
Surfacing this to @scheglov 
BTW thank you for fixing the other related bugs. I appreciate it a lot!

**Comment by scheglov** [comment]
This issue was fixed by 0ab26a77519b316372a2249af1c853e3645d4296.
